src/evaluate.rs

```diff
@@ -17,14 +17,17 @@ use rayon::prelude::*;
 
 #[cfg(not(feature = "parallel"))]
 use crate::rayon;
-use crate::{atomicmin::AtomicMin, deflate, filters::RowFilter, png::PngImage, Deadline, PngError};
+use crate::{
+    atomicmin::AtomicMin, deflate, filters::FilterStrategy, png::PngImage, Deadline, PngError,
+};
 
 pub(crate) struct Candidate {
     pub image: Arc<PngImage>,
     pub data: Vec<u8>,
     pub data_is_compressed: bool,
     pub estimated_output_size: usize,
-    pub filter: RowFilter,
+    pub filter: FilterStrategy,
+    pub filter_used: FilterStrategy,
```

Comment:
These two are confusing. I should come up with clear names and add descriptive comments.

Reply:
Perhaps a better option would be to just make `Predefined` take a string description of where it came from.

For a radically different idea: Maybe the filters can just be different implementations of a trait, with all the heuristic logic contained within the implementation. This could even allow library implementers to create their own strategies. I'm not sure I'm prepared to do such a refactor right now though.

Reply:
Okay, I've reverted to the two properties on Candidate, as we need to retain the input filter for APNG frames.